x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetTopNFunctionsAction.java

```diff
@@ -0,0 +1,171 @@
+        StopWatch watch = new StopWatch("getTopNFunctionsAction");
+        client.execute(GetStackTracesAction.INSTANCE, request, ActionListener.wrap(searchResponse -> {
+            long responseStart = System.nanoTime();
+            StopWatch processingWatch = new StopWatch("Processing response");
+            GetTopNFunctionsResponse topNFunctionsResponse = buildTopNFunctions(searchResponse);
+            log.debug(() -> watch.report() + " " + processingWatch.report());
+            listener.onResponse(topNFunctionsResponse);
+        }, listener::onFailure));
+    }
+
+    static GetTopNFunctionsResponse buildTopNFunctions(GetStackTracesResponse response) {
+        TopNFunctionsBuilder builder = new TopNFunctionsBuilder(response.getSamplingRate());
+        if (response.getTotalFrames() == 0) {
+            return builder.build();
+        }
+
+        for (Map.Entry<String, StackTrace> st : response.getStackTraces().entrySet()) {
+            Set<String> frameGroupsPerStackTrace = new HashSet<String>();
+            String stackTraceId = st.getKey();
+            StackTrace stackTrace = st.getValue();
+            long samples = response.getStackTraceEvents().getOrDefault(stackTraceId, 0L);
+            builder.addTotalCount(samples);
+
+            int frameCount = stackTrace.frameIds.size();
+            for (int i = 0; i < frameCount; i++) {
+                String frameId = stackTrace.frameIds.get(i);
+                String fileId = stackTrace.fileIds.get(i);
+                Integer frameType = stackTrace.typeIds.get(i);
+                Integer addressOrLine = stackTrace.addressOrLines.get(i);
+                StackFrame stackFrame = response.getStackFrames().getOrDefault(frameId, EMPTY_STACKFRAME);
+                String executable = response.getExecutables().getOrDefault(fileId, "");
+
+                for (Frame frame : stackFrame.frames()) {
+                    String frameGroupId = FrameGroupID.create(fileId, addressOrLine, executable, frame.fileName(), frame.functionName());
+                    if (builder.setCurrentTopNFunction(frameGroupId) == false) {
+                        builder.addTopNFunction(
+                            frameGroupId,
+                            new StackFrameMetadata(
+                                frameId,
+                                fileId,
+                                frameType,
+                                frame.inline(),
+                                addressOrLine,
+                                frame.functionName(),
+                                frame.functionOffset(),
+                                frame.fileName(),
+                                frame.lineNumber(),
+                                executable
+                            )
+                        );
+                    }
+                    if (frameGroupsPerStackTrace.contains(frameGroupId) == false) {
```

Comment:
I *think* I understand why this guard is here but IMHO this warrants a longer comment explaining it (I believe it's there because we want to avoid counting recursive function calls multiple times in the same stack trace)

Reply:
You're right about the reason behind the guard. I added comments to clarify.